From 30a4d32153bbc6d2eb90c88e01f5aeda0ad8da20 Mon Sep 17 00:00:00 2001
From: Ladi Prosek <lprosek@redhat.com>
Date: Mon, 13 Feb 2017 16:27:59 +0000
Subject: [PATCH] vgabios: Reorder video modes to work around a Windows bug

RH-Author: Ladi Prosek <lprosek@redhat.com>
Message-id: <1487003279-21425-2-git-send-email-lprosek@redhat.com>
Patchwork-id: 73784
O-Subject: [RHEL-6.9 vgabios PATCH 1/1] vgabios: Reorder video modes to work around a Windows bug
Bugzilla: 1421574
RH-Acked-by: Gerd Hoffmann <kraxel@redhat.com>
RH-Acked-by: Laszlo Ersek <lersek@redhat.com>
RH-Acked-by: Yan Vugenfirer <yvugenfi@redhat.com>

Windows Server 2016 and Windows 10 RS1 come with a bug in its blue screen
of death rendering logic which prevents it from generating crash dumps.

The bug does not manifest if Windows sees a suitable 32 bpp video mode
before a suitable 24 bpp video mode in the list of modes returned from
vgabios. This commit moves all 32 bpp modes to the front of the list to
make sure that this is always the case.

Upstream patch:
https://www.coreboot.org/pipermail/seabios/2016-October/010963.html

There are valid concerns upstream about the breaking nature of the fix
but for the limited set of operating systems supported by RHEL/RHEV we
can easily verify that they are unaffected. So as things stand now, this
is a downstream-only patch which will be reverted in the near future;
the exact time will depend on Windows 10 RS2 schedule and other factors.
The goal is to make sure that our customers running Windows 10 VMs can
generate crash dumps.

Signed-off-by: Ladi Prosek <lprosek@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1421574
Brew: https://brewweb.engineering.redhat.com/brew/taskinfo?taskID=12535405
Signed-off-by: Danilo C. L. de Paula <ddepaula@redhat.com>
---
 vbetables-gen.c | 34 ++++++++++++++++++----------------
 1 file changed, 18 insertions(+), 16 deletions(-)

diff --git a/vbetables-gen.c b/vbetables-gen.c
index 96ac85a..a605628 100644
--- a/vbetables-gen.c
+++ b/vbetables-gen.c
@@ -12,6 +12,23 @@ typedef struct {
 } ModeInfo;
 
 ModeInfo modes[] = {
+      /* BOCHS/PLE, 86 'own' mode numbers - 32 bpp */
+{ 320, 200, 32                        , 0x140},
+{ 640, 400, 32                        , 0x141},
+{ 640, 480, 32                        , 0x142},
+{ 800, 600, 32                        , 0x143},
+{ 1024, 768, 32                       , 0x144},
+{ 1280, 1024, 32                      , 0x145},
+{ 1600, 1200, 32                      , 0x147},
+{ 1152, 864, 32                      , 0x14c},
+{ 1280, 800, 32                      , 0x17a},
+{ 1280, 960, 32                      , 0x17d},
+{ 1440, 900, 32                      , 0x180},
+{ 1400, 1050, 32                     , 0x183},
+{ 1680, 1050, 32                     , 0x186},
+{ 1920, 1200, 32                     , 0x189},
+{ 2560, 1600, 32                     , 0x18c},
+
     /* standard VESA modes */
 { 640, 400, 8                          , 0x100},
 { 640, 480, 8                          , 0x101},
@@ -41,41 +58,26 @@ ModeInfo modes[] = {
 { 1600, 1200, 16                      , 0x11E},
 { 1600, 1200, 24                      , 0x11F},
 
-      /* BOCHS/PLE, 86 'own' mode numbers */
-{ 320, 200, 32                        , 0x140},
-{ 640, 400, 32                        , 0x141},
-{ 640, 480, 32                        , 0x142},
-{ 800, 600, 32                        , 0x143},
-{ 1024, 768, 32                       , 0x144},
-{ 1280, 1024, 32                      , 0x145},
+      /* BOCHS/PLE, 86 'own' mode numbers - 8, 15, 16, and 24 bpp */
 { 320, 200, 8                           , 0x146},
-{ 1600, 1200, 32                      , 0x147},
 { 1152, 864, 8                      , 0x148},
 { 1152, 864, 15                      , 0x149},
 { 1152, 864, 16                      , 0x14a},
 { 1152, 864, 24                      , 0x14b},
-{ 1152, 864, 32                      , 0x14c},
 { 1280, 800, 16                      , 0x178},
 { 1280, 800, 24                      , 0x179},
-{ 1280, 800, 32                      , 0x17a},
 { 1280, 960, 16                      , 0x17b},
 { 1280, 960, 24                      , 0x17c},
-{ 1280, 960, 32                      , 0x17d},
 { 1440, 900, 16                      , 0x17e},
 { 1440, 900, 24                      , 0x17f},
-{ 1440, 900, 32                      , 0x180},
 { 1400, 1050, 16                     , 0x181},
 { 1400, 1050, 24                     , 0x182},
-{ 1400, 1050, 32                     , 0x183},
 { 1680, 1050, 16                     , 0x184},
 { 1680, 1050, 24                     , 0x185},
-{ 1680, 1050, 32                     , 0x186},
 { 1920, 1200, 16                     , 0x187},
 { 1920, 1200, 24                     , 0x188},
-{ 1920, 1200, 32                     , 0x189},
 { 2560, 1600, 16                     , 0x18a},
 { 2560, 1600, 24                     , 0x18b},
-{ 2560, 1600, 32                     , 0x18c},
 { 0, },
 };
 
-- 
1.8.3.1

